src/control/server/storage/bdev/backend_vmd.go

```diff
@@ -0,0 +1,236 @@
+//
+// (C) Copyright 2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package bdev
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/logging"
+	"github.com/daos-stack/daos/src/control/server/storage"
+)
+
+// findPciAddrsWithDomain returns controllers that match the input prefix in the
+// domain component of their PCI address.
+func findPciAddrsWithDomain(inCtrlrs storage.NvmeControllers, prefix string) ([]string, error) {
+	var outPciAddrs []string
+
+	for _, ctrlr := range inCtrlrs {
+		domain, _, _, _, err := common.ParsePCIAddress(ctrlr.PciAddr)
+		if err != nil {
+			return nil, err
+		}
+		if fmt.Sprintf("%x", domain) == prefix {
+			outPciAddrs = append(outPciAddrs, ctrlr.PciAddr)
+		}
+	}
+
+	return outPciAddrs, nil
+}
+
+// substVMDAddrs replaces VMD PCI addresses in input device list with the
+// PCI addresses of the backing devices behind the VMD.
+//
+// Select any PCI addresses that have the compressed VMD address BDF as backing
+// address domain.
+//
+// Return new device list with PCI addresses of devices behind the VMD.
+func substVMDAddrs(inPCIAddrs []string, foundCtrlrs storage.NvmeControllers) ([]string, error) {
+	if len(foundCtrlrs) == 0 {
+		return nil, nil
+	}
+
+	var outPciAddrs []string
+	for _, dev := range inPCIAddrs {
+		_, b, d, f, err := common.ParsePCIAddress(dev)
+		if err != nil {
+			return nil, err
+		}
+		matchDevs, err := findPciAddrsWithDomain(foundCtrlrs,
+			fmt.Sprintf("%02x%02x%02x", b, d, f))
+		if err != nil {
+			return nil, err
+		}
+		if len(matchDevs) == 0 {
+			outPciAddrs = append(outPciAddrs, dev)
+			continue
+		}
+		outPciAddrs = append(outPciAddrs, matchDevs...)
+	}
+
+	return outPciAddrs, nil
+}
+
+// substituteVMDAddresses wraps around substVMDAddrs and takes a BdevScanResponse
+// reference along with a logger.
+func substituteVMDAddresses(log logging.Logger, inPCIAddrs []string, bdevCache *storage.BdevScanResponse) ([]string, error) {
+	if bdevCache == nil || len(bdevCache.Controllers) == 0 {
+		log.Debugf("no bdev cache to find vmd backing devices (devs: %v)", inPCIAddrs)
+		return nil, nil
+	}
+
+	msg := fmt.Sprintf("vmd detected, processing addresses (input %v, existing %v)",
+		inPCIAddrs, bdevCache.Controllers)
+
+	dl, err := substVMDAddrs(inPCIAddrs, bdevCache.Controllers)
+	if err != nil {
+		return nil, errors.Wrapf(err, msg)
+	}
+	log.Debugf("%s: new %s", msg, dl)
+
+	return dl, nil
+}
+
+// detectVMD returns whether VMD devices have been found and a slice of VMD
+// PCI addresses if found.
+func detectVMD() ([]string, error) {
+	// Check available VMD devices with command:
+	// "$lspci | grep  -i -E "201d | Volume Management Device"
+	lspciCmd := exec.Command("lspci")
+	vmdCmd := exec.Command("grep", "-i", "-E", "201d|Volume Management Device")
```

Comment:
We should only grep for "Volume Management Device" now, since the 201d pci ID is only true for Skylake/Cascade Lake and has changed in ICX systems. Should simplify the logic a bit.